design-proposals/e2e-framework.md

```diff
@@ -0,0 +1,92 @@
+
+It is also a symptom of a kind of non-independence of the tests from each other:
+test A does not care about cleaning the cluster, so much as it knows that there will be
+the setup of test B will do it.
+
+Last but not least, each test file must specify the cleanup and setup function on its own, and in most cases
+is something like this:
+```go
+BeforeEach(func() {
+  tests.BeforeTestCleanup()
+  
+  virtClient, err = kubecli.GetKubevirtClient()
+  util.PanicOnError(err)
+  //other test specific setup
+})
+```
+
+The `Framework`  can help resolve these issues by supporting tests in setting up and cleaning up the environment
+in which they will operate. This will provide a separation of duties in the various phases of a test.
+This will make it easier to script the tests since you will not have to worry about setup and cleanup
+of the environment. It will also define a test architecture that is easier for new members to understand.
+
+## Goals
+
+- Support common operations.
+- Increase e2e test stabilization.
+- Minimizing the possibility of leftovers after a test run.
+
+## Non Goals
+- Refactoring all the e2e tests at once
+
+# Proposal
+
+## Definition of Personas
+
+- E2E test authors.
+
+## User Stories
+
+As an e2e test author:
+
+I want to create a new e2e test file ensuring that there will be no leftover that may influence subsequent testing.
+
+## Repos
+Kubevirt/kubevirt
+
+## Design
+The `Framework` should be similar to Kubernetes' [Framework](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/framework/framework.go).
+Phases of a test suite are:
+1. **Environment setup**: prepares the cluster by installing namespaces, resources, etc..
+2. **Test setup**: prepares the cluster with changes that are specific to a group of tests.
+3. **Test execution**
+4. **Test teardown**: must restore the environment to the previous state of the setup.
+5. **Environment teardown**:  must restore the environment to the previous state of the environment setup.
+
+| Phase                | Ginkgo function           | Repetition                                    |
+|----------------------|:--------------------------|:----------------------------------------------|
+| Environment setup    | SynchronizedBeforeSuite() | Once at the beginning of the whole test suite |
+| Test setup           | BeforeEach()              | Before each test execution                    |
+| Test execution       |                           |                                               |
+| Test teardown        | AfterEach()               | After each test execution                     |
+| Environment teardown | SynchronizedAfterSuite()  | Once at the end of the whole test suite       |
+
+The `Framework` supports steps 2 and 4, freeing the test files from all the common steps of preparing and
+cleaning the environment. Note that steps 2 and 4 will not disappear in the test files, but will contain
+only the logic related to that specific set of tests.
```

Comment:
Maybe also explain the cleaning mechanism as well?
I really liked the "CleanableResource" idea from your original PR, maybe it can be discussed here as well